### PR TITLE
Working draft of lad-4 flight code

### PR DESCRIPTION
**Pull request by nlautrette**

### CODE REVIEW

File: lib/Barometer/Barometer.cpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
+/*
+  IMU.cpp - A c++ library to interface with the Arduino BMP3XX Barometer.
+  Created by Nolan Lautrette, March 21, 2021.
+*/
```
Comment (fazerlicourice7) [inline lib/Barometer/Barometer.cpp]: fix comment

File: lib/GPS/GPS.cpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+/**
+ * Define this in advance. Need to be agreed on by everyone.
+ */
+void GPS::readAuxilliaryData(float *data) {
+  _gps.read();
+  data[0] = _gps.altitude;
+  data[1] = _gps.speed;
+  data[2] = _gps.angle;
+  data[3] = _gps.satellites;
+  data[4] = -1;
```
Comment (fazerlicourice7) [inline lib/GPS/GPS.cpp]: change this order based on packet def. Add fix data (is that just num satellites? or was there extra). 
And update packet def to include angle

File: lib/apogeeDetection/apogeeDetection.cpp [hunk 3]
```diff
@@ -5,35 +5,40 @@
...
+		_R = new MatrixXd(m, m);
 		*_R << altitudeVar , 0,
 		 			0,     accelVar;
 
-		_z = new VectorXd(2);
+		_z = new VectorXd(m);
 
-		kalmanfilter = new Kalman(3, *_F, 2, 1, *_H, MatrixXd::Zero(3,1), *_Q, *_R);
+		_x = new VectorXd(n);
+		*_x << initAlt, initAcc, 0;
```
Comment (fazerlicourice7) [inline lib/apogeeDetection/apogeeDetection.cpp]: this should be `initAlt, 0, initAcc;`

File: platformio.ini [hunk 4]
```diff
@@ -24,16 +24,16 @@ lib_deps =
 
 [env:E1_coldflow]
 src_filter = -<*> +<E1_coldflow/>
-build_flags = -DHAL=teensy ; -DDEBUG
```
Comment (fazerlicourice7) [inline platformio.ini]: undo these changes. we want to be able to easily uncomment, or comment out the DEBUG flag

File: src/LAD4_1/config.h [hunk 5]
```diff
@@ -1,72 +1,34 @@
...
-const uint8_t numSensors = 4;
-struct sensorInfo **sensors;
+const uint8_t numSensors = 7;
+sensorInfo *sensors;
 
 const float batteryMonitorShuntR = 0.002; // ohms
 const float batteryMonitorMaxExpectedCurrent = 10; // amps
 
-#define DROGUE_PIN 9
-#define MAIN_CHUTE_PIN 10
```
Comment (fazerlicourice7) [inline src/LAD4_1/config.h:33]: this should stay. make it a `const uint8_t` though.

File: src/LAD4_1/config.h [hunk 6]
```diff
@@ -1,72 +1,34 @@
...
+    sensors[0] = {"IMU Acceleration",   FLIGHT_BRAIN_ADDR, 14, 1};
+    sensors[1] = {"IMU Orientation", FLIGHT_BRAIN_ADDR, 15, 1};
+    sensors[2] = {"Barometer", FLIGHT_BRAIN_ADDR, 13, 1};
+    sensors[3] = {"Battery Stats", FLIGHT_BRAIN_ADDR, 2, 3};
+    sensors[4] = {"GPS Lat Long",  FLIGHT_BRAIN_ADDR, 11, 3};
+    sensors[5] = {"GPS AUX",      FLIGHT_BRAIN_ADDR, 12, 3};
+    sensors[6] = {"Number Packets Sent", FLIGHT_BRAIN_ADDR, 5, 10};
 
-    pinMode(DROGUE_PIN, OUTPUT);
-    pinMode(MAIN_CHUTE_PIN, OUTPUT);
```
Comment (fazerlicourice7) [inline src/LAD4_1/config.h]: This stays

File: src/LAD4_1/config.h [hunk 7]
```diff
@@ -1,72 +1,34 @@
...
+    //debug("Initializing sensors", DEBUG);
+    sensors = new sensorInfo[numSensors];
+    // the ordering in this array defines order of operation, not id
+    sensors[0] = {"IMU Acceleration",   FLIGHT_BRAIN_ADDR, 14, 1};
+    sensors[1] = {"IMU Orientation", FLIGHT_BRAIN_ADDR, 15, 1};
+    sensors[2] = {"Barometer", FLIGHT_BRAIN_ADDR, 13, 1};
+    sensors[3] = {"Battery Stats", FLIGHT_BRAIN_ADDR, 2, 3};
+    sensors[4] = {"GPS Lat Long",  FLIGHT_BRAIN_ADDR, 11, 3};
+    sensors[5] = {"GPS AUX",      FLIGHT_BRAIN_ADDR, 12, 3};
+    sensors[6] = {"Number Packets Sent", FLIGHT_BRAIN_ADDR, 5, 10};
```
Comment (fazerlicourice7) [inline src/LAD4_1/config.h:45]: you will need to add a `sensor` that checks apogee. We Also need to add a sensor for main chute deployment at 300m or something.

File: src/LAD4_1/lad4_1.cpp [hunk 8]
```diff
@@ -1,5 +1,176 @@
...
+
+void sensorReadFunc(int id);
+
+IMU _imu;
+Barometer bmp;
+GPS gps(GPS_Serial);
+ApogeeDetection detector;
+
+double acc_z;
+double altitude;
```
Comment (fazerlicourice7) [inline src/LAD4_1/lad4_1.cpp]: why are these two here. we shouldn't need any of this here. all of that should go inside the apogee detection library.

File: src/LAD4_1/lad4_1.cpp [hunk 9]
```diff
@@ -1,5 +1,176 @@
...
+
+  bmp.init(&Wire);
+
+  gps.init();
+
+  // Initializes initial state _x in Kalman filter to the alt and acc of the rocket at setup
+  sensorReadFunc(0);
+  acc_z = farrbconvert.sensorReadings[2];
+  sensorReadFunc(4);
+  altitude = farrbconvert.sensorReadings[0];
```
Comment (fazerlicourice7) [inline src/LAD4_1/lad4_1.cpp]: there is no case 0 or 4 in `sensorReadFunc`

File: src/LAD4_1/lad4_1.cpp [hunk 10]
```diff
@@ -1,5 +1,176 @@
...
+
+  double altVar = 0.5;
+  double accVar = 0.5;
```
Comment (fazerlicourice7) [inline src/LAD4_1/lad4_1.cpp]: these two go in config as `const double`

File: src/LAD4_1/lad4_1.cpp [hunk 11]
```diff
@@ -1,5 +1,176 @@
...
+  detector = ApogeeDetection(20e-3, altVar, accVar, altitude, acc_z);
```
Comment (fazerlicourice7) [inline src/LAD4_1/lad4_1.cpp]: If we can't construct this statically, let's modify the apogee detection library to be like the IMU and barometer libraries above w/ a default constructor and a separate init func that we can pass all of this to.
20e-3 should also be a variable defined in config.

File: src/LAD4_1/lad4_1.cpp [hunk 12]
```diff
@@ -1,5 +1,176 @@
...
+    sensor = &sensors[j];
+    sensorReadFunc(sensor->id);
+    if(sensor->id == 0) {
+      acc_z = farrbconvert.sensorReadings[0];
+    }
+    else if(sensor->id == 2) {
+      altitude = farrbconvert.sensorReadings[2];
+      if(detector.atApogee(altitude, acc_z)) {
+        //deloy chute
+      }
```
Comment (fazerlicourice7) [inline src/LAD4_1/lad4_1.cpp]: You need to define sensors in config. and move any of this logic to the `sensorReadFunc` defined below.
Ig, you'll need the altitude and acc_z variables you defined above. so ignore that comment.

File: src/LAD4_1/lad4_1.cpp [hunk 13]
```diff
@@ -1,5 +1,176 @@
...
+void sensorReadFunc(int id) {
+  switch (id) {
+    case 14:
+      _imu.readAccelerationData(farrbconvert.sensorReadings);
+      break;
+    case 15:
+      _imu.readOrientationData(farrbconvert.sensorReadings);
+      break;
+    case 13:
+      bmp.readAllData(farrbconvert.sensorReadings);
```
Comment (fazerlicourice7) [inline src/LAD4_1/lad4_1.cpp]: update altitude data here.

File: src/LAD4_1/lad4_1.cpp [hunk 14]
```diff
@@ -1,5 +1,176 @@
...
+    write_to_SD(packet.c_str(), file_name);
+  }
+  delay(10);
+
+}
+
+void sensorReadFunc(int id) {
+  switch (id) {
+    case 14:
+      _imu.readAccelerationData(farrbconvert.sensorReadings);
```
Comment (fazerlicourice7) [inline src/LAD4_1/lad4_1.cpp]: update accel data here.

File: src/LAD4_1/lad4_1.cpp [hunk 15]
```diff
@@ -1,5 +1,176 @@
...
+        RFSerial.println(packet);
+    #endif
+    write_to_SD(packet.c_str(), file_name);
+  }
+  delay(10);
+
+}
+
+void sensorReadFunc(int id) {
+  switch (id) {
```
Comment (fazerlicourice7) [inline src/LAD4_1/lad4_1.cpp:130]: update the switch to be in case order 2, 5, 11, 12, 13, 14, 15....

**Review by fazerlicourice7 — CHANGES_REQUESTED** [review]

File: src/LAD4_1/config.h (hunk 7 above)
Comment (nlautrette) [reply src/LAD4_1/config.h:45]: @ssalcedo00

File: lib/apogeeDetection/apogeeDetection.cpp (hunk 3 above)
Comment (nlautrette) [reply lib/apogeeDetection/apogeeDetection.cpp]: @ssalcedo00

File: src/LAD4_1/lad4_1.cpp (hunk 8 above)
Comment (nlautrette) [reply src/LAD4_1/lad4_1.cpp]: @ssalcedo00

File: platformio.ini (hunk 4 above)
Comment (nlautrette) [reply platformio.ini]: is this supposed to be grayed out @fazerlicourice7

Comment (fazerlicourice7) [reply platformio.ini]: leave the `-DDEBUG` as a comment (the `;` is the comment identifier in`.ini` files)

Comment (salsquared) [reply platformio.ini]: is this resolved?

File: lib/recovery/recovery.h [hunk 16]
```diff
@@ -19,8 +20,8 @@ namespace Recovery {
     drogue_chute_state = 0;
     main_chute_state = 0;
 
-    digitalWrite(drogue_pin, drogue_chute_state);
-    digitalWrite(main_pin, main_chute_state);
+    digitalWrite(DROGUE_PIN, drogue_chute_state);
+    digitalWrite(MAIN_CHUTE_PIN, main_chute_state);
```
Comment (fazerlicourice7) [inline lib/recovery/recovery.h]: nope. update the init to take in two pin numbers, set the value of `drogue_pin` and `main_pin` to the values that were passed in . Use those two vars locally

File: lib/apogeeDetection/apogeeDetection.cpp [hunk 17]
```diff
@@ -57,5 +65,16 @@ bool ApogeeDetection::atApogee(double altitude, double accel_z) {
 	return false;
 }
 
+void ApogeeDetection::updateAlt(float* data) {
+	altitude = data[0];
+}
+
+void ApogeeDetection::updateAcc(float* data) {
+	acc_z = data[2];
+}
```
Comment (fazerlicourice7) [inline lib/apogeeDetection/apogeeDetection.cpp]: nice.
send in just a single float though. 
it should be called like `updateAlt(farrbconvert.sensorReadings[0])`

File: src/LAD4_1/lad4_1.cpp [hunk 18]
```diff
@@ -1,5 +1,181 @@
...
+  debug("Initializing Libraries");
+  batteryMonitor::init(&Wire, batteryMonitorShuntR, batteryMonitorMaxExpectedCurrent);
+  _imu.init(&Wire);
+  bmp.init(&Wire);
+  gps.init();
+
+  bmp.readAllData(farrbconvert.sensorReadings);
+  initAcc_z = farrbconvert.sensorReadings[2];
+  _imu.readAccelerationData(farrbconvert.sensorReadings);
+  initAlt = farrbconvert.sensorReadings[0];
```
Comment (fazerlicourice7) [inline src/LAD4_1/lad4_1.cpp]: check this order.

File: src/LAD4_1/lad4_1.cpp [hunk 19]
```diff
@@ -1,5 +1,181 @@
...
+  if(290 <= alt && alt <= 310) 
+    return true;
+  else 
+    return false;
+}
+
+void recoveryPacket() {
+  Recovery::getAllStates(farrbconvert.sensorReadings);
+  packet = make_packet(10, false);
+  Serial.println(packet);
```
Comment (fazerlicourice7) [inline src/LAD4_1/lad4_1.cpp:185]: also send to rf, and write to sd

File: src/LAD4_1/config.h [hunk 20]
```diff
@@ -1,70 +1,47 @@
...
```
Comment (fazerlicourice7) [inline src/LAD4_1/config.h:45]: add sensor for recovery pack

File: src/LAD4_1/lad4_1.cpp [hunk 21]
```diff
@@ -1,5 +1,181 @@
...
+        Recovery::releaseMainChute();
+        recoveryPacket();
+      }
+      break;
+    case 14:
+      _imu.readAccelerationData(farrbconvert.sensorReadings);
+      if(detector.atApogee()) {
+        passedApogee = true;
+        Recovery::releaseDrogueChute();
+        recoveryPacket();
```
Comment (fazerlicourice7) [inline src/LAD4_1/lad4_1.cpp:161]: you'll need to close the chute after a certain delay go with 10ms for now, Andy is testing, we may update that number.

File: src/LAD4_1/lad4_1.cpp [hunk 22]
```diff
@@ -1,5 +1,181 @@
...
+      bmp.readAllData(farrbconvert.sensorReadings);
+      detector.updateAlt(farrbconvert.sensorReadings);
+      if(passedApogee && aroundMainDeploy()) {
+        Recovery::releaseMainChute();
+        recoveryPacket();
+      }
+      break;
+    case 14:
+      _imu.readAccelerationData(farrbconvert.sensorReadings);
+      if(detector.atApogee()) {
```
Comment (fazerlicourice7) [inline src/LAD4_1/lad4_1.cpp]: remember we only need to do this if we're past main engine cut off. You added that to the python lib, let's add that here too

**Review by fazerlicourice7 — CHANGES_REQUESTED** [review]
looking much better. Still some work before it's ready. 
Last stretch!

File: platformio.ini (hunk 4 above)
Comment (fazerlicourice7) [reply platformio.ini]: yeah

File: src/LAD4_1/config.h (hunk 20 above)
Comment (salsquared) [reply src/LAD4_1/config.h:45]: i thought we werent gonna add a sensor for recovery

File: lib/GPS/GPS.cpp [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+GPS::GPS(SPIClass *theSPI, int8_t cspin) {
+  _gps = Adafruit_GPS(theSPI, cspin);
+  commMethod = 4;
+  init();
+}
+
+void GPS::init() {
+    _gps.begin(9600);
+
+    _gps.sendCommand(PMTK_SET_NMEA_OUTPUT_RMCONLY);
```
Comment (fazerlicourice7) [inline lib/GPS/GPS.cpp]: Change this to `PMTK_SET_NMEA__OUTPUT_RMCGGA`

File: src/LAD4_1/config.h (hunk 20 above)
Comment (fazerlicourice7) [reply src/LAD4_1/config.h:45]: we should send it periodically (very slowly) 
but also send it right after you do the recovery commands

Comment (salsquared) [reply src/LAD4_1/config.h:45]: every 10, 20 cycles good?

File: src/LAD4_1/lad4_1.cpp (hunk 21 above)
Comment (salsquared) [reply src/LAD4_1/lad4_1.cpp:161]: what is there to close lol we're using black powder. or is it gonna be sending a signal until i close it?

**Comment by fazerlicourice7** [comment]
make delay like 500 ms for chute
